File: frontend/lib/mappers/cashflow-mapper.ts
/**
 * Cashflow Mapper - Stage 4 Cashflow Truth Workspace
 *
 * Transforms backend DTOs to ViewModels for the Cashflow Truth Workspace.
 * This is the ONLY location where DTO to ViewModel mapping occurs.
 *
 * Architecture Flow: Backend → API → DTO → Mapper → ViewModel
 */

import type {
  CashflowViewModel,
  CashflowTrendViewModel,
  CashflowMonthlyViewModel,
  CashflowCategoryViewModel,
  CashflowTransactionViewModel,
  CashflowInsightViewModel,
  CashflowEvidenceChainViewModel,
  CashflowFiltersViewModel,
  CashflowNavigationViewModel,
} from '../../types/cashflow-view-model';

// ===== DTO Types (from backend) =====

type CashflowTrendDirection = 'up' | 'down' | 'flat';

interface CashflowTrendDTO {
  direction: CashflowTrendDirection;
  percentage_change: number;
  period: string;
  volatility_score: number;
}

interface CashflowMonthlyDTO {
  month: string;
  income_paise: number;
  expenses_paise: number;
  net_paise: number;
  transaction_count: number;
}

interface CashflowCategoryDTO {
  category_id: string;
  category_name: string;
  amount_paise: number;
  percentage: number;
  transaction_count: number;
}

interface CashflowTransactionDTO {
  id: string;
  date: string;
  description: string;
  amount_paise: number;
  category: string;
  merchant?: string;
}

type CashflowInsightType = 'positive' | 'warning' | 'info' | 'alert';
type CashflowInsightSeverity = 'low' | 'medium' | 'high';

interface CashflowInsightDTO {
  type: CashflowInsightType;
  severity: CashflowInsightSeverity;
  message: string;
  action_url?: string;
}

interface CashflowEvidenceItemDTO {
  type: string;
  summary: string;
  source: string;
  confidence?: number;
}

interface CashflowCalculationStepDTO {
  name: string;
  description: string;
  inputs: Record<string, unknown>;
  outputs: Record<string, unknown>;
}

interface CashflowEvidenceChainDTO {
  summary: string;
  evidence: CashflowEvidenceItemDTO[];
  calculation_steps: CashflowCalculationStepDTO[];
  source_references: string[];
  confidence_score: number;
}

interface CashflowSummaryDTO {
  total_income_paise: number;
  total_expenses_paise: number;
  net_cashflow_paise: number;
  transaction_count: number;
  trend?: CashflowTrendDTO;
  insights: CashflowInsightDTO[];
  evidence_chain?: CashflowEvidenceChainDTO;
}

/**
 * Cashflow Mapper Interface
 * Defines the contract for DTO to ViewModel mapping
 */
export interface ICashflowMapper {
  /**
   * Map a single Cashflow DTO to ViewModel
   */
  mapCashflowDTO(dto: CashflowSummaryDTO): CashflowViewModel;

  /**
   * Map monthly summaries DTOs to ViewModels
   */
  mapMonthlySummaries(dtos: CashflowMonthlyDTO[]): CashflowMonthlyViewModel[];

  /**
   * Map category breakdowns DTOs to ViewModels
   */
  mapCategoryBreakdowns(dtos: CashflowCategoryDTO[]): CashflowCategoryViewModel[];

  /**
   * Map transactions DTOs to ViewModels
   */
  mapTransactions(dtos: CashflowTransactionDTO[]): CashflowTransactionViewModel[];

  /**
   * Map trend DTO to ViewModel
   */
  mapTrendDTO(dto: CashflowTrendDTO | null | undefined): CashflowTrendViewModel | undefined;

  /**
   * Map insights DTOs to ViewModels
   */
  mapInsights(dtos: CashflowInsightDTO[]): CashflowInsightViewModel[];

  /**
   * Map evidence chain DTO to ViewModel
   */
  mapEvidenceChain(
    dto: CashflowEvidenceChainDTO | null | undefined
  ): CashflowEvidenceChainViewModel | undefined;
}

/**
 * Cashflow Mapper Implementation
 * Transforms backend cashflow data to ViewModels
 */
export class CashflowMapper implements ICashflowMapper {
  /**
   * Map a single Cashflow DTO to ViewModel
   * @param dto - Cashflow data from API
   * @returns CashflowViewModel for presentation
   */
  mapCashflowDTO(dto: CashflowSummaryDTO): CashflowViewModel {
    return {
      total_income_paise: dto.total_income_paise,
      total_expenses_paise: dto.total_expenses_paise,
      net_cashflow_paise: dto.net_cashflow_paise,
      transaction_count: dto.transaction_count,
      trend: this.mapTrendDTO(dto.trend),
      monthly: [],
      categories: [],
      transactions: [],
      insights: this.mapInsights(dto.insights),
      evidence_chain: this.mapEvidenceChain(dto.evidence_chain),
      filters: this.createDefaultFilters(),
      navigation: this.createDefaultNavigation(),
    };
  }

  /**
   * Map monthly summaries DTOs to ViewModels
   */
  mapMonthlySummaries(dtos: CashflowMonthlyDTO[]): CashflowMonthlyViewModel[] {
    if (!dtos || dtos.length === 0) {
      return [];
    }
    return dtos.map((dto) => ({
      month: dto.month,
      income_paise: dto.income_paise,
      expenses_paise: dto.expenses_paise,
      net_paise: dto.net_paise,
      transaction_count: dto.transaction_count,
    }));
  }

  /**
   * Map category breakdowns DTOs to ViewModels
   */
  mapCategoryBreakdowns(dtos: CashflowCategoryDTO[]): CashflowCategoryViewModel[] {
    if (!dtos || dtos.length === 0) {
      return [];
    }
    return dtos.map((dto) => ({
      category_id: dto.category_id,
      category_name: dto.category_name,
      amount_paise: dto.amount_paise,
      percentage: dto.percentage,
      transaction_count: dto.transaction_count,
    }));
  }

  /**
   * Map transactions DTOs to ViewModels
   */
  mapTransactions(dtos: CashflowTransactionDTO[]): CashflowTransactionViewModel[] {
    if (!dtos || dtos.length === 0) {
      return [];
    }
    return dtos.map((dto) => ({
      id: dto.id,
      date: dto.date,
      description: dto.description,
      amount_paise: dto.amount_paise,
      category: dto.category,
      merchant: dto.merchant,
    }));
  }

  /**
   * Map trend DTO to ViewModel
   */
  mapTrendDTO(
    dto: CashflowTrendDTO | null | undefined
  ): CashflowTrendViewModel | undefined {
    if (!dto) {
      return undefined;
    }
    return {
      direction: dto.direction,
      percentage_change: dto.percentage_change,
      period: dto.period,
      volatility_score: dto.volatility_score,
    };
  }

  /**
   * Map insights DTOs to ViewModels
   */
  mapInsights(dtos: CashflowInsightDTO[]): CashflowInsightViewModel[] {
    if (!dtos || dtos.length === 0) {
      return [];
    }
    return dtos.map((dto) => ({
      type: dto.type,
      severity: dto.severity,
      message: dto.message,
      action_url: dto.action_url,
    }));
  }

  /**
   * Map evidence chain DTO to ViewModel
   */
  mapEvidenceChain(
    dto: CashflowEvidenceChainDTO | null | undefined
  ): CashflowEvidenceChainViewModel | undefined {
    if (!dto) {
      return undefined;
    }
    return {
      summary: dto.summary,
      evidence: dto.evidence.map((item) => ({
        type: item.type,
        summary: item.summary,
        source: item.source,
        confidence: item.confidence,
      })),
      calculation_steps: dto.calculation_steps.map((step) => ({
        name: step.name,
        description: step.description,
        inputs: step.inputs,
        outputs: step.outputs,
      })),
      source_references: dto.source_references,
      confidence_score: dto.confidence_score,
    };
  }

  /**
   * Create default filters
   */
  private createDefaultFilters(): CashflowFiltersViewModel {
    return {
      date_range: undefined,
      categories: undefined,
      merchants: undefined,
      amount_range: undefined,
    };
  }

  /**
   * Create default navigation
   */
  private createDefaultNavigation(): CashflowNavigationViewModel {
    return {
      deep_link: '/cashflow',
      cross_references: {
        accounts: '/accounts',
        transactions: '/transactions',
      },
    };
  }
}

// Export singleton instance
export const cashflowMapper = new CashflowMapper();

File: backend/src/core/mappers/cashflow_mapper.py
"""Cashflow Mapper

Transforms cashflow domain objects into CashflowDTO instances.
This is the ONLY location where cashflow API responses are constructed.
"""

from typing import Any

from src.core.dtos.cashflow_dto import (
    CashflowCategoryDTO,
    CashflowInsightDTO,
    CashflowMonthlyDTO,
    CashflowSummaryDTO,
    CashflowTransactionDTO,
    CashflowTrendDTO,
)


class CashflowMapper:
    """Mapper for cashflow domain objects to DTOs."""

    @staticmethod
    def to_summary_dto(cashflow_data: dict[str, Any]) -> CashflowSummaryDTO:
        """Convert cashflow data to CashflowSummaryDTO."""
        return CashflowSummaryDTO(
            total_income_paise=cashflow_data.get("total_income_paise", 0),
            total_expenses_paise=cashflow_data.get("total_expenses_paise", 0),
            net_cashflow_paise=cashflow_data.get("net_cashflow_paise", 0),
            transaction_count=cashflow_data.get("transaction_count", 0),
            trend=CashflowMapper._to_trend_dto(cashflow_data.get("trend")),
            insights=CashflowMapper._to_insights(cashflow_data.get("insights", [])),
            evidence_chain=cashflow_data.get("evidence_chain"),
        )

    @staticmethod
    def _to_trend_dto(trend_data: dict[str, Any] | None) -> CashflowTrendDTO | None:
        """Convert trend data to CashflowTrendDTO."""
        if not trend_data:
            return None
        return CashflowTrendDTO(
            direction=trend_data.get("direction", "flat"),
            percentage_change=trend_data.get("percentage_change", 0.0),
            period=trend_data.get("period", "1M"),
            volatility_score=trend_data.get("volatility_score", 0.0),
        )

    @staticmethod
    def _to_insights(insights: list[dict[str, Any]]) -> list[CashflowInsightDTO]:
        """Convert insights to CashflowInsightDTO list."""
        return [
            CashflowInsightDTO(
                type=insight.get("type", "info"),
                severity=insight.get("severity", "medium"),
                message=insight.get("message", ""),
                action_url=insight.get("action_url"),
            )
            for insight in insights
        ]

    @staticmethod
    def to_monthly_dto(monthly_data: dict[str, Any]) -> CashflowMonthlyDTO:
        """Convert monthly cashflow data to CashflowMonthlyDTO."""
        return CashflowMonthlyDTO(
            month=monthly_data.get("month", ""),
            income_paise=monthly_data.get("income_paise", 0),
            expenses_paise=monthly_data.get("expenses_paise", 0),
            net_paise=monthly_data.get("net_paise", 0),
            transaction_count=monthly_data.get("transaction_count", 0),
        )

    @staticmethod
    def to_category_dto(category_data: dict[str, Any]) -> CashflowCategoryDTO:
        """Convert category data to CashflowCategoryDTO."""
        return CashflowCategoryDTO(
            category_id=category_data.get("category_id", ""),
            category_name=category_data.get("category_name", "Unknown"),
            amount_paise=category_data.get("amount_paise", 0),
            percentage=category_data.get("percentage", 0.0),
            transaction_count=category_data.get("transaction_count", 0),
        )

    @staticmethod
    def to_transaction_dto(transaction_data: dict[str, Any]) -> CashflowTransactionDTO:
        """Convert transaction data to CashflowTransactionDTO."""
        return CashflowTransactionDTO(
            id=transaction_data.get("id", ""),
            date=transaction_data.get("date", ""),
            description=transaction_data.get("description", ""),
            amount_paise=transaction_data.get("amount_paise", 0),
            category=transaction_data.get("category", "Uncategorized"),
            merchant=transaction_data.get("merchant"),
        )
